lib/clamav/client.rb

```diff
@@ -24,30 +24,131 @@
 require "clamav/wrappers/null_termination_wrapper"
 
 module ClamAV
+  DEBUG = ENV.fetch('DEBUG', '0').strip == '1'
+
   class Client
-    def initialize(connection = default_connection)
-      @connection = connection
-      connection.establish_connection
+    class Error < StandardError; end
+    class ConnectionError < Error; end
+    class ConnectTimeoutError < ConnectionError; end
+
+    attr_writer :unix_socket
+    attr_writer :tcp_host
+    attr_writer :tcp_port
+    attr_writer :connect_timeout
+    attr_writer :connection
+
+    def initialize(*args)
+      args.each do |arg|
+        case arg
+        when Connection
+          @connection = arg
+        when Hash
+          arg.each do |attr, value|
+            send("#{attr}=", value)
```

Comment:
I think i'd rather avoid `send` and implement an abstraction to the two different types of sockets.